In [136]:
import numpy as np
import pandas as pd
from itertools import chain, groupby
from collections import defaultdict
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.cross_validation import cross_val_score
from sklearn.feature_extraction.text import HashingVectorizer

## Подготовка данных (urls)

Я попытался учесть количество раз, с которым url встречается в браузерной истории пользователя.

In [2]:
urls_train_df = pd.read_csv('kaggle_data/url_domain_train', header=None, delimiter='\t')
urls_train_df.columns = ['id', 'url', 'count']
urls_train_df.head()

id                  url  count
0  000000014B60815F65B38258011B6C01  login.rutracker.org      1
1  000000014B60815F65B38258011B6C01        rutracker.org      4
2  000000014C03DA2A47AC433A0C755201   admin.tour-spb.net      1
3  000000014C03DA2A47AC433A0C755201            czinfo.ru      1
4  000000014C03DA2A47AC433A0C755201       forumsostav.ru      1

In [3]:
urls_train_df = pd.DataFrame(urls_train_df.groupby('id').apply(lambda df: list(zip(df['url'], df['count']))))
urls_train_df['id'] = urls_train_df.index
urls_train_df.index = range(len(urls_train_df))
urls_train_df.columns = ['urls', 'id']
urls_train_df.head()

urls  \
0  [(id.rambler.ru, 1), (mail.rambler.ru, 1), (r0...   
1  [(1prime.ru, 1), (autorambler.ru, 1), (chellak...   
2                            [(bosch-korolev.ru, 1)]   
3  [(aptekanizkihcen.ua, 1), (colady.ru, 1), (gor...   
4  [(astrorok.ru, 1), (diets.ru, 1), (edaplus.inf...   

                                 id  
0  000000013CB5719C0000A2C90002C101  
1  00000001442BE24000001B7D00F50801  
2  00000001448580F800003F1B31FB0901  
3  0000000145BDB2FF000157971645E901  
4  000000014602771F0000DB9359714C01

In [4]:
age_train_df = pd.read_csv('kaggle_data/age_profile_train', header=None, delimiter='\t')
age_train_df.columns = ['id', 'age']
print("Max age: {}".format(max(age_train_df['age'])))
print("Min age: {}".format(min(age_train_df['age'])))
age_train_df.head()

Max age: 99
Min age: 0


id  age
0  000000013CB5719C0000A2C90002C101   53
1  00000001442BE24000001B7D00F50801   48
2  00000001448580F800003F1B31FB0901   28
3  0000000145BDB2FF000157971645E901   44
4  000000014602771F0000DB9359714C01   48

In [50]:
train_df = urls_train_df.merge(age_train_df, on='id', how='left')
print(len(train_df))
train_df[train_df['age'] == 58]

118603


urls  \
33      [(45fss.ru, 2), (fss.ru, 2), (kurgan.spravker....   
117                                [(mail.rambler.ru, 1)]   
197                 [(freehouse.ru, 2), (topnomer.ru, 1)]   
290                                [(mail.rambler.ru, 1)]   
328     [(auto.vesti.ru, 1), (filmpro.ru, 1), (inforea...   
546     [(1prime.ru, 1), (assessor.ru, 1), (base.consu...   
560     [(assets.adobedtm.com, 1), (championat.com, 3)...   
710     [(ecson.ru, 1), (mail.rambler.ru, 19), (o-voyn...   
733     [(1tv.ru, 2), (belaruspartisan.org, 1), (dieta...   
832     [(bom-bom.ru, 1), (ffclub.ru, 1), (horoscopes....   
846                  [(mail.rambler.ru, 5), (rubo.ru, 1)]   
1009                               [(mail.rambler.ru, 1)]   
1101    [(03.ru, 1), (1000cxem.com, 1), (1tv.ru, 1), (...   
1177    [(docs.cntd.ru, 1), (mail.rambler.ru, 12), (mo...   
1304    [(id.rambler.ru, 1), (mail.rambler.ru, 13), (r...   
1372             [(lacywear.ru, 1), (mail.rambler.ru, 7)]   
1511    [(ishariki.ru, 1), (mail.rambler.ru, 1), (news...   
1516    [(altravita-ivf.ru, 1), (arbuzspb.ru, 1), (bn....   
1525    [(24stoma.ru, 1), (2do2go.ru, 1), (allinform.r...   
1594    [(images.rambler.ru, 1), (mail.rambler.ru, 12)...   
1662    [(101.ru, 1), (1366x768.ru, 1), (agent.mail.ru...   
1676    [(1001-dollar.com, 1), (aif.ru, 1), (approblem...   
1882    [(agroserver.ru, 1), (aif.ru, 1), (autorambler...   
1915    [(aeroflot.ru, 1), (id.rambler.ru, 1), (m.aero...   
1968    [(7days.ru, 1), (download.drweb.ru, 1), (drweb...   
2024    [(7days.ru, 1), (aif.ru, 1), (autonews.ru, 1),...   
2106    [(cekatop.ru, 1), (images.rambler.ru, 1), (int...   
2122    [(assessor.ru, 1), (base.consultant.ru, 1), (c...   
2124    [(abisoft.ru, 1), (aif.ru, 1), (autorambler.ru...   
2264    [(aif.ru, 1), (autorambler.ru, 1), (avtodispet...   
...                                                   ...   
113454  [(amerricana.ru, 1), (besumka.livejournal.com,...   
113752  [(10minecraft.ru, 5), (archive.espec.ws, 1), (...   
114229  [(mail.rambler.ru, 11), (ukrainians-world.org....   
114344  [(aif.ru, 3), (autorambler.ru, 2), (davydov-in...   
114349  [(adv.msk.ru, 1), (consultant.ru, 1), (law-raa...   
114773  [(aup.ru, 1), (cfin.ru, 1), (citforum.ru, 1), ...   
114910  [(catchenglish.ru, 1), (englishtopics.net, 1),...   
115062  [(ag.ru, 1), (amanmarket.com, 1), (bezprogramm...   
115096  [(auto.drom.ru, 1), (barabinsk.tiu.ru, 1), (ch...   
115263  [(b-port.com, 1), (base.garant.ru, 1), (consul...   
115312  [(0564.ua, 1), (dgex-studio.ru, 1), (forum.vgd...   
115316  [(0552online.com, 14), (101trava.ru, 1), (an-k...   
115834  [(100pereezdov.ru, 1), (enter.unicredit.ru, 2)...   
116626  [(all-profy.ru, 1), (cmclub.ru, 1), (confy.ru,...   
116646  [(beats4you.ru, 1), (business-vending.ru, 1), ...   
116743  [(bsphere.ru, 1), (homeopatica.ru, 1), (homeos...   
116746  [(7nebo-otel.ru, 1), (mail.rambler.ru, 1), (na...   
116899  [(bestsmoking.ru, 1), (bigpicture.ru, 1), (cre...   
117227  [(expert.ru, 1), (kino-teatr.ru, 1), (kommersa...   
117228  [(aquarists.ru, 1), (feltdoll.blogspot.ru, 1),...   
117230  [(kramola.info, 5), (mail.rambler.ru, 2), (new...   
117310  [(afve.org, 1), (cosmo.ru, 1), (family.rambler...   
117577        [(mail-pda.rambler.ru, 1), (rambler.ru, 1)]   
117667  [(autorambler.ru, 1), (mail.rambler.ru, 1), (r...   
117739  [(championat.com, 1), (cosmo.ru, 1), (expert.r...   
117955            [(mail.rambler.ru, 1), (rambler.ru, 1)]   
118152  [(5minshop.com, 1), (dialog.ua, 1), (forbes.ru...   
118301  [(actuallno.com, 1), (aif.ru, 1), (mail.ramble...   
118591  [(kino-teatr.ru, 1), (knijky.ru, 8), (zak.depo...   
118599  [(aif.ru, 1), (autochel.ru, 5), (chelyabinsk.r...   

                                      id  age  
33      000000014993E4FE0AF232A20A374C01   58  
117     000000014BB63F59B332AD2502C15401   58  
197     000000014D00FADA5629475402FAE401   58  
290     000000014E12FCD1517F2108000CF301   58  
328     000000014E22A4870B372E

## Снижение размерности

Здесь я попытался реализовать то, что было описано на семинаре.
Для каждого $i \in \{0, \ldots, 9\}$ я считаю $P(age \in [10 i, 10 (i + 1)) | url)$. В результате у меня получается 10 словарей с условными вероятностями для каждого url и для каждого отрезка возраста.

In [6]:
len(train_df)

118603

In [7]:
def calc_probs(train_df, from_age, to_age):
    total_visits = defaultdict(float)
    cond_visits = defaultdict(float)
    for domains, age in zip(train_df.urls.values, train_df.age.values):
        for domain, count in domains:
            if age in range(from_age, to_age):
                cond_visits[domain] += count
            total_visits[domain] += count
    for key in cond_visits:
        cond_visits[key] /= total_visits[key]
    return cond_visits

conditionals = []
for i in range(10):
    print(i)
    conditionals.append(calc_probs(train_df, 10 * i, 10 * (i + 1)))

0
1
2
3
4
5
6
7
8
9


Далее, вектор фичей, которые отвечают за пользователя, строится так: для каждого отрезка возраста строится гистограмма из 100 бакетов, и все они сливаются. Получается вектор длины 1000. К сожалению некоторые компоненты этих векторов получаются зависимыми, но я так и не смог придумать, как учитывать одновременно много промежутков возростов и сделать так, чтобы фичи были независимы. Можно, конечно, сделать $P(age \in [0, 25] | url)$, но так получается, что мы слишком сильно теряем информации из входных данных.

In [115]:
def transform(urls):
    x = np.zeros((1000,))
    for url, count in urls:
        for i in range(10):
            if url in conditionals[i]:
                probability = conditionals[i][url]
                x[100 * i + int((probability - 1e-6) * 100)] += count
            
    return x / np.linalg.norm(x)

In [20]:
y = train_df.age.values
X = []
for val in train_df.urls.values:
    X.append(transform(val))

In [21]:
len(y)

118603

In [22]:
X = np.array(X)
X.shape

(118603, 1000)

In [69]:
X[1]

array([ 0.01667593,  0.38910512,  0.04446916,  0.        ,  0.        ,
        0.        ,  0.        ,  0.00555864,  0.        ,  0.        ,
        0.00555864,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.  

## Обучение модели

Здесь я написал небольшую cross_validation, я попробовал Ridge (линейная модель с L2-регуляризацией), Lasso (L1 резуляризация, в теории должна лучше работать).

In [107]:
regressions = [Lasso, Ridge]
for alpha in [1e-4, 1e-3, 1e-2, 1e-1, 1, 10]:
    for reg in regressions:
        r = reg(alpha=alpha)
        print(reg, alpha)
        scores = (-cross_val_score(r, X, y, scoring='mean_squared_error'))
        print(scores)
        print(sum(scores) / 3)
        print('----------')

/Users/astepanov/anaconda3/lib/python3.5/site-packages/sklearn/linear_model/coordinate_descent.py:466: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations
  ConvergenceWarning)
/Users/astepanov/anaconda3/lib/python3.5/site-packages/sklearn/linear_model/coordinate_descent.py:466: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations
  ConvergenceWarning)


<class 'sklearn.linear_model.coordinate_descent.Lasso'> 0.0001
[ 121.83068989  130.3887478    96.73304434]
116.317494011
----------
<class 'sklearn.linear_model.ridge.Ridge'> 0.0001
[ 122.53057224  130.99959883  101.18705273]
118.239074601
----------
<class 'sklearn.linear_model.coordinate_descent.Lasso'> 0.001
[ 125.70604072  132.07278338  101.14903567]
119.642619923
----------
<class 'sklearn.linear_model.ridge.Ridge'> 0.001
[ 122.10582529  130.77051804   99.1455297 ]
117.340624344
----------
<class 'sklearn.linear_model.coordinate_descent.Lasso'> 0.01
[ 136.52766534  138.04645292  112.26152955]
128.945215939
----------
<class 'sklearn.linear_model.ridge.Ridge'> 0.01
[ 121.4773583   130.57298842   97.62278189]
116.557709536
----------
<class 'sklearn.linear_model.coordinate_descent.Lasso'> 0.1
[ 155.44611844  150.9030644   133.95268106]
146.767287965
----------
<class 'sklearn.linear_model.ridge.Ridge'> 0.1
[ 121.7787172   130.47703769   97.42165158]
116.559135488
----------
<class '

/Users/astepanov/anaconda3/lib/python3.5/site-packages/sklearn/linear_model/coordinate_descent.py:466: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations
  ConvergenceWarning)


Для сравнения ElasticNet (композиция L2 и L1 резуляризаций)

In [113]:
r = ElasticNet(alpha=0.01, l1_ratio=0.01)
scores = (-cross_val_score(r, X, y, scoring='mean_squared_error'))
print(scores)
print(sum(scores) / 3)

[ 148.16604999  145.71214886  127.74198659]
140.540061811


## Отправка Решения

Было решено использовать Ridge(alpha=0.1). К сожалению, не смотря на хороший скор на cross_validation, это решение набрало на kaggle 148.

In [114]:
reg = Ridge(alpha=0.1)
reg.fit(X, y)

Ridge(alpha=0.1, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [123]:
urls_test_df = pd.read_csv('kaggle_data/url_domain_test', header=None, delimiter='\t')
urls_test_df.columns = ['id', 'url', 'count']
urls_test_df.head()

id                          url  count
0  0000000151004FF4ADD746DA10685A01                    afisha.ru      2
1  0000000151004FF4ADD746DA10685A01                       aif.ru      1
2  0000000151004FF4ADD746DA10685A01  aimfar.solution.weborama.fr      1
3  0000000151004FF4ADD746DA10685A01                  alkotest.ru      1
4  0000000151004FF4ADD746DA10685A01                 aptekamos.ru      1

In [124]:
urls_test_df = pd.DataFrame(urls_test_df.groupby('id').apply(lambda df: list(zip(df['url'], df['count']))))
urls_test_df['id'] = urls_test_df.index
urls_test_df.index = range(len(urls_test_df))
urls_test_df.columns = ['urls', 'id']
urls_test_df.head()

urls  \
0  [(1000bankov.ru, 1), (1tv.ru, 8), (4put.ru, 1)...   
1  [(autorambler.ru, 1), (bilettorg.ru, 1), (dsol...   
2              [(photosight.ru, 6), (rambler.ru, 6)]   
3  [(base.consultant.ru, 1), (dogovor-obrazets.ru...   
4  [(assessor.ru, 2), (audit-it.ru, 1), (base.gar...   

                                 id  
0  000000014A02348E701552980349FF01  
1  000000014A10EA183BF8594A0B2AB201  
2  000000014A4FE5C33A929D4C26943601  
3  000000014B7BB9957784A9BC0AC9F401  
4  000000014C7749F896D82C2B01E8B801

In [125]:
X_test = []
for val in urls_test_df.urls.values:
    X_test.append(transform(val))
X_test = np.array(X_test)

In [126]:
X_test.shape

(19974, 1000)

In [127]:
y_pred = reg.predict(X_test)

In [128]:
y_pred

array([ 39.06696015,  42.0904562 ,  51.14819456, ...,  37.54280448,
        37.19906284,  38.12296796])

In [129]:
urls_test_df['age'] = y_pred

In [130]:
urls_test_df = urls_test_df[['id', 'age']]
urls_test_df.columns = ['Id', 'age']

In [131]:
urls_test_df.head()

Id        age
0  000000014A02348E701552980349FF01  39.066960
1  000000014A10EA183BF8594A0B2AB201  42.090456
2  000000014A4FE5C33A929D4C26943601  51.148195
3  000000014B7BB9957784A9BC0AC9F401  40.113299
4  000000014C7749F896D82C2B01E8B801  37.109740

In [132]:
random_sol = pd.read_csv('kaggle_data/random_solution.csv')
miss_idx = set(random_sol.Id.values) - set(urls_test_df.Id.values)
miss_df = pd.DataFrame(list(zip(list(miss_idx), np.ones(len(miss_idx)))))
miss_df.columns = ['Id', 'age']

In [133]:
urls_test_df = urls_test_df.append(miss_df, ignore_index=True)

In [134]:
urls_test_df.to_csv('kaggle_data/solution.csv', index=False)

In [135]:
!wc -l kaggle_data/solution.csv

   19980 kaggle_data/solution.csv


FIN.